lib/resolve-package-path.js

```diff
@@ -0,0 +1,170 @@
+    if (stat.isFile() || stat.isFIFO()) {
+      realPath = fs.realpathSync(filePath);
+    } else {
+      realPath = null; 
+    }
+  } catch (e) {
+    if (
+      e !== null &&
+      typeof e === 'object' &&
+      (e.code === 'ENOENT' || e.code === 'ENOTDIR')
+    ) {
+      realPath = null; // NO PATH FOUND
+    } else {
+      throw e;
+    }
+  }
+
+  cache.set(filePath, realPath);
+
+  return realPath;
+}
+
+/*
+ * Given a package 'name' and starting directory, resolve to a real (existing) file path.
+ *
+ * Do it similar to how it is done in resolve.sync() - travel up the directory hierarchy,
+ * attaching 'node-modules' to each directory and seeing if the directory exists and
+ * has the relevant 'package.json' file we're searching for. It is *much* faster than
+ * resolve.sync(), because we don't test that the requested name is a directory.
+ * This routine assumes that it is only called when we don't already have
+ * the cached entry.
+ * 
+ * NOTE: it is valid for 'name' to be an absolute or relative path and 'dir' to 
+ * be null or undefined. If 'dir' is undefined, it will be set to __dirname, or
+ * ignored if name actually is an absolute path.
+ *
+ * @param realFilePathCache the cache containing the real paths corresponding to
+ * various file paths (which may or may not be already resolved). This is only used
+ * if passed through to _getRealFilePath.
+ * @param name the 'name' of the module, i.e. x in require(x), but with
+ *   '/package.json' on the end. It is NOT a directory (so we don't have to do the
+ *   directory checks that resolve.sync does).
+ *   NOTE: because this is an internal routine, for speed it does not check 
+ *   that '/package.json' is actually the end of the name.
+ * @param dir the directory (MUST BE non-empty, and valid) to start from, appending the name to the
+ *   directory and checking that the file exists. Go up the directory hierarchy from there.
+ *   if name is itself an absolute path, 
+ *
+ * @result the path to the actual package.json file that's found, or null if not.
+ */
+function _findPackagePath(realFilePathCache, name, dir) {
+
+  var h = heimdall.start('hashForDep:_findPackagePath');
+
+  try {
+    console.log('_findPackagePath for name ' + name + ', dir ' + dir);
+
+    var fsRoot = pathRoot(dir);
+
+    console.log('root = ' + fsRoot);
+
+    var currPath = dir;
+
+    // force the currPath to end with node_modules, as that's how we'll
+    // go up the directory tree.
+    if (path.basename(currPath)  !== 'node_modules') {
+      currPath = currPath + path.sep + 'node_modules';
+    }
+
+    while (currPath !== fsRoot) {
+      var filePath = path.join(currPath, name);  // node_modules/name
+
+      var realPath = _getRealFilePath(realFilePathCache, filePath);
+
+      if (realPath) {
+        return realPath;
+      }
+
+      // pop to next higher node_modules. This assumes that
+      // node_modules/node_modules is not possible.
+      currPath = path.dirname(currPath);
```

Comment:
I don't believe this is correct as per my earlier recommendation, the double "pop" of the current directory should only occur if the `currPath` ends in node_modules.

see: https://github.com/stefanpenner/hash-for-dep/pull/45/files#r254023431


Also, node_modules can occur at ever directory level not just the first one. Given:

`findPackagePath(cache, 'apple', '/foo/bar/baz/node_modules/')`

I would expect the following locations be considered as a location for `apple`

```
'/foo/node_modules/bar/orange/baz/node_modules/apple'
'/foo/bar/orange/node_modules/apple'
'/foo/bar/node_modules/apple'
'/foo/node_modules/apple'
'/node_modules/apple'
```

Reply:
yep, that's where it should go. The change I put in is wrong here. Will fix today.